File: bin/dbt-plot-sar-blockdev.r
#!/bin/sh
#
# This file is released under the terms of the Artistic License.
# Please see the file LICENSE, included in this package, for details.
#
# Copyright (C) 2014      Mark Wong
#               2014      2ndQuadrant, Ltd.
#

INPUTFILE=$1
OUTPUTDIR=$2

if [ $# -ne 2 ]; then
	echo "Create a plot the rate of transactions."
	echo "usage: `basename $0` <sar-blockdev.csv> <output directory>"
	echo
	echo "    <sar-blockdev.csv> - full path to the sar-blockdev.csv file"
	echo "    <output directory> - location to write output files"
	echo
	echo "Will attempt to create <output directory> if it does not exist."
	exit 1
fi

R --slave --no-save << __EOF__
df <- read.csv("$INPUTFILE", sep = ";", header=T)
starttime = df[1,]\$timestamp
df\$timestamp <- ceiling((df\$timestamp - starttime) / 60)
devices <- sort(unique(df\$DEV))
count <- length(devices)
color <- rainbow(count)
pch <- seq.int(1, count)

# Convert kilobytes to megabytes, without renaming the column...
df\$rkB.s <- df\$rkB.s / 1024
df\$wkB.s <- df\$wkB.s / 1024

# Plot all device utilization on a single chart.

df.all <- subset(df, df\$DEV == devices[1])
bitmap("${OUTPUTDIR}/sar-blockdev-util.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df.all\$timestamp, df.all\$X.util,
     ylim=c(0, 100), type = "b", col = color[1],
     main="Block Device Utilization",
     xlab="Elapsed Time (minutes)", ylab="% Utilization")

for (i in 2:count) {
  df.all <- subset(df, df\$DEV == devices[i])
  points(df.all\$timestamp, df.all\$X.util, type = "b", pch = pch[i],
         col=color[i])
}

legend('topright', sapply(devices, as.character), pch=pch, col=color)

grid(col="gray")
invisible(dev.off())

# Plot all device tps on a single chart.

df.all <- subset(df, df\$DEV == devices[1])
bitmap("${OUTPUTDIR}/sar-blockdev-tps.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df.all\$timestamp, df.all\$tps,
     ylim=c(0, max(df\$tps)), type = "b", col = color[1],
     main="Transfers per Second",
     xlab="Elapsed Time (minutes)", ylab="")

for (i in 2:count) {
  df.all <- subset(df, df\$DEV == devices[i])
  points(df.all\$timestamp, df.all\$tps, type = "b", pch = pch[i],
         col=color[i])
}

legend('topright', sapply(devices, as.character), pch=pch, col=color)

grid(col="gray")
invisible(dev.off())

df.all <- subset(df, df\$DEV == devices[1])
bitmap("${OUTPUTDIR}/sar-blockdev-await.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df.all\$timestamp, df.all\$await,
     ylim=c(0, max(df\$await)), type = "b", col = color[1],
     main="Average Wait",
     xlab="Elapsed Time (minutes)", ylab="Milliseconds")

for (i in 2:count) {
  df.all <- subset(df, df\$DEV == devices[i])
  points(df.all\$timestamp, df.all\$await, type = "b", pch = pch[i],
         col=color[i])
}

legend('topright', sapply(devices, as.character), pch=pch, col=color)

grid(col="gray")
invisible(dev.off())

df.all <- subset(df, df\$DEV == devices[1])
bitmap("${OUTPUTDIR}/sar-blockdev-avgqu.sz.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df.all\$timestamp, df.all\$aqu.sz,
     ylim=c(0, max(df\$aqu.sz)), type = "b", col = color[1],
     main="Average Queue Length",
     xlab="Elapsed Time (minutes)", ylab="Length")

for (i in 2:count) {
  df.all <- subset(df, df\$DEV == devices[i])
  points(df.all\$timestamp, df.all\$aqu.sz, type = "b", pch = pch[i],
         col=color[i])
}

legend('topright', sapply(devices, as.character), pch=pch, col=color)

grid(col="gray")
invisible(dev.off())

df.all <- subset(df, df\$DEV == devices[1])
bitmap("${OUTPUTDIR}/sar-blockdev-avgrq.sz.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df.all\$timestamp, df.all\$areq.sz,
     ylim=c(0, max(df\$areq.sz)), type = "b", col = color[1],
     main="Average Request Size",
     xlab="Elapsed Time (minutes)", ylab="Sectors")

for (i in 2:count) {
  df.all <- subset(df, df\$DEV == devices[i])
  points(df.all\$timestamp, df.all\$areq.sz, type = "b", pch = pch[i],
         col=color[i])
}

legend('topright', sapply(devices, as.character), pch=pch, col=color)

grid(col="gray")
invisible(dev.off())

# Plot megabytes read/write per device

label <- c('Reads', 'Writes')
color <- rainbow(2)
pch <- c(1, 2)
for (dev in devices) {
  df.all <- subset(df, df\$DEV == dev)
  bitmap(paste("${OUTPUTDIR}/sar-blockdev-", dev, "-rw.png", sep = ""),
         type="png16m", units="px", width=1280, height=800, res=150, taa=4,
         gaa=4)
  plot(df.all\$timestamp, df.all\$rkB.s,
       ylim=c(0, max(max(df.all\$rkB.s), max(df.all\$wkB.s))),
       type = "b", col = color[1],
       main=paste("Block Device ", dev, " Read/Write", sep = ""),
       xlab="Elapsed Time (minutes)", ylab="MegaBytes/s")
  points(df.all\$timestamp, df.all\$wkB.s, type = "b", pch = pch[2],
         col=color[2])

  legend('topright', label, pch=pch, col=color)

  grid(col="gray")
dev.off()

}
__EOF__

File: bin/dbt-plot-sar-blockdev.jl
#!/bin/sh
#
# This file is released under the terms of the Artistic License.
# Please see the file LICENSE, included in this package, for details.
#
# Copyright The DBT Tools Authors
#

#=
INPUTFILE=$1
OUTPUTDIR=$2

if [ $# -ne 2 ]; then
	echo "Create a plot the rate of transactions."
	echo "usage: $(basename "$0") <sar-blockdev.csv> <output directory>"
	echo
	echo "    <sar-blockdev.csv> - full path to the sar-blockdev.csv file"
	echo "    <output directory> - location to write output files"
	echo
	echo "Will attempt to create <output directory> if it does not exist."
	exit 1
fi

exec julia --color=no --startup-file=no "$0" "$@"
=#

using CSV
using DataFrames
using VegaLite

function load(params)
    df = CSV.File(params["filename"]) |> DataFrame

    # Convert timestamp to elapsed time in minutes, in-place.

    start_time = minimum(df.timestamp)
    df = df[!, Not(["# hostname", "interval"])]
    transform!(
            df,
            :timestamp => ByRow(x -> (x - start_time) / 60) => :timestamp;
    )

    return df
end

function plot(df, params)
    p = df |>
        @vlplot(
            title="Block Device Utilization",
            width=1200,
            height=800,
            mark={
                type="line",
                point=true,
            },
            x={
                :timestamp,
                title="Elapsed Time (minutes)",
            },
            y={
                field="%util",
                title="% Utilization",
                scale={
                    domain=[0, 100],
                },
            },
            color={
                :DEV,
                title="Block Device",
            },
        )
    filename = params["outputdir"] * "/sar-blockdev-util.png"
    save(filename, p)

    p = df |>
        @vlplot(
            title="Transfer per Second",
            width=1200,
            height=800,
            mark={
                type="line",
                point=true,
            },
            x={
                :timestamp,
                title="Elapsed Time (minutes)",
            },
            y={
                field=:tps,
                title="",
            },
            color={
                :DEV,
                title="Block Device",
            },
        )
    filename = params["outputdir"] * "/sar-blockdev-tps.png"
    save(filename, p)

    p = df |>
        @vlplot(
            title="Average Wait",
            width=1200,
            height=800,
            mark={
                type="line",
                point=true,
            },
            x={
                :timestamp,
                title="Elapsed Time (minutes)",
            },
            y={
                field=:await,
                title="Milliseconds",
            },
            color={
                :DEV,
                title="Block Device",
            },
        )
    filename = params["outputdir"] * "/sar-blockdev-await.png"
    save(filename, p)

    p = df |>
        @vlplot(
            title="Average Queue Length",
            width=1200,
            height=800,
            mark={
                type="line",
                point=true,
            },
            x={
                :timestamp,
                title="Elapsed Time (minutes)",
            },
            y={
                field="aqu-sz",
                title="",
            },
            color={
                :DEV,
                title="Block Device",
            },
        )
    filename = params["outputdir"] * "/sar-blockdev-avgqu.png"
    save(filename, p)

    p = df |>
        @vlplot(
            title="Average Request Size",
            width=1200,
            height=800,
            mark={
                type="line",
                point=true,
            },
            x={
                :timestamp,
                title="Elapsed Time (minutes)",
            },
            y={
                field="areq-sz",
                title="Sectors",
            },
            color={
                :DEV,
                title="Block Device",
            },
        )
    filename = params["outputdir"] * "/sar-blockdev-areq.png"
    save(filename, p)

    devs = sort(unique(df.DEV))
    for dev in devs
        sdf = df[df.DEV .== dev, ["timestamp", "rkB/s", "wkB/s"]]
        plot_dev(dev, sdf, params)
    end
end

function plot_dev(dev, df, params)
    rdf = df[!, ["timestamp", "rkB/s"]]
    rename!(rdf, ["timestamp", "mB/s"])
    insertcols!(rdf, :label => "Reads")

    wdf = df[!, ["timestamp", "wkB/s"]]
    rename!(wdf, ["timestamp", "mB/s"])
    insertcols!(wdf, :label => "Writes")
    rdf = vcat(rdf, wdf)
    transform!(rdf, "mB/s" => ByRow(x -> x / 1024.0) => "mB/s")

    p = rdf |>
        @vlplot(
            title="Block Device " * dev * " Read/Write",
            width=1200,
            height=800,
            mark={
                type="line",
                point=true,
            },
            x={
                :timestamp,
                title="Elapsed Time (minutes)",
            },
            y={
                field="mB/s",
                title="MegaBytes/s",
            },
            color={
                :label,
                title="",
            },
        )
    filename = params["outputdir"] * "/sar-blockdev-" * dev * "-rw.png"
    save(filename, p)
end

function main()
    params = Dict(
            "filename" => ARGS[1],
            "outputdir" => ARGS[2],
    )

    df = load(params)
    plot(df, params)
end

main()
